staging/src/k8s.io/cloud-provider/controllers/nodelifecycle/node_lifecycle_controller.go

```diff
@@ -207,7 +208,24 @@ func shutdownInCloudProvider(ctx context.Context, cloud cloudprovider.Interface,
 		return false, errors.New("cloud provider does not support instances")
 	}
 
-	shutdown, err := instances.InstanceShutdownByProviderID(ctx, node.Spec.ProviderID)
+	providerID := node.Spec.ProviderID
+	if providerID == "" {
+		var err error
+		providerID, err = instances.InstanceID(ctx, types.NodeName(node.Name))
```

Comment:
I faced the same issue and created a PR for it (already closed it).
Maybe we have to change it here as well, the instance ID is no provider ID and has no provider prefix.
Don't know if it's necessary as the `InstanceShutdownByProviderID` is stripping the provider prefix anyway. But maybe it's worth of adding it, analogous to this in PR https://github.com/kubernetes/kubernetes/pull/121383
https://github.com/kubernetes/kubernetes/blob/8c0af3ee5ba52eefcc61740fa3d94206ea3f4d3c/staging/src/k8s.io/cloud-provider/controllers/nodelifecycle/node_lifecycle_controller.go#L269

Reply:
How do you want to handle that? I can see what you mean about the diff between provider ID and instance ID, that is valid. At the time when I created this patch I had no idea, but at the same time something along the way makes the assignment I do here a working one (like you suggest, something strips off the prefix later?)

I am happy to add prefix here but I am also happy to keep it as it is given that you tested the patch and also for my environment it fixed the issues I had.

Reply:
Totally fine for me to keep it as is. Maybe we can create a follow issue/PR that fixes all occurrences.

Reply:
If we want more fine-grained control over fetching the instance ID, we could just add another `getProviderId` method that takes in the `instances.InstanceID` result. Very small change. Seems kind of wrong to build an invalid provider ID when we know that's what we're doing and it's a minor change. But I do agree that fixes all of the other occurrences makes sense as a follow up.

Reply:
Just take a look at my PR https://github.com/kubernetes/kubernetes/pull/121383, I introduced a method `getProviderID` :)

Reply:
Ok as it's not a big deal I created a function `getProviderID` similarly to the current `ensureNodeExistsByProviderID` and `shutdownInCloudProvider`. It will do what we discussed here, prepending the provider name so we are building a correct provider ID.

I would appreciate a look regarding the instanceV2 handling inside `getProviderID`. Today it is not used because `shutdownInCloudProvider` handles instanceV2 interface separately and will never call `getProviderID` but I still want to handle it in case someone somewhere start using `getProviderID` (because why not if it's convenient)

Reply:
nit: `providerID, err = instances.InstanceID(ctx, types.NodeName(node.Name))` shouldn't `providerID` be `instanceID` in this case?

Reply:
Mhm, of course that makes sense. I thought I could piggyback on having only single variable used all around but it's not a good idea